### DISCUSSION TITLE
Actions including docker build on self hosted runner behind proxy

### DISCUSSION THREAD

**Discussion by macoto-takahashi** [community/community#55601] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi. I have self-hosted runners behind our web proxy.
When I try to use actions including docker build, such as _actions-hub/kubectl@master_ on my workflow,  , _Build actions-hub/kubectl@master_  job fails because of lack of web accessibility in the docker build process. Of cause docker host has http_proxy env, so the docker host can pull the base image, but cannot build because of lack of docker build --build-arg option.

I know it works if I fork _actions-hub/kubectl@master_ then add ENV line on the Dockerfile, but there are any other oprions?

```
<SNIP`>

  #7 [3/3] RUN chmod +x /entrypoint.sh &&     apk add --no-cache --update openssl curl ca-certificates &&     curl -L https://storage.googleapis.com/kubernetes-release/release/v1.27.1/bin/linux/amd64/kubectl -o /usr/local/bin/kubectl &&     chmod +x /usr/local/bin/kubectl &&     rm -rf /var/cache/apk/*
  #7 sha256:a9407e2da77d2563c0954dce7efca99df8a13d9d807b34e1c52793d6da809dfb
  #7 0.203 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz
  #7 130.2 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz
  #7 130.2 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/main: network error (check Internet connection and firewall)
  #7 261.3 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/community: network error (check Internet connection and firewall)
  #7 261.3 ERROR: unable to select packages:
  #7 261.3   ca-certificates (no such package):
  #7 261.3     required by: world[ca-certificates]
  #7 261.3   curl (no such package):
  #7 261.3     required by: world[curl]
  #7 261.3   openssl (no such package):
  #7 261.3     required by: world[openssl]
  #7 ERROR: process "/bin/sh -c chmod +x /entrypoint.sh &&     apk add --no-cache --update openssl curl ca-certificates &&     curl -L https://storage.googleapis.com/kubernetes-release/release/$KUBE_VERSION/bin/linux/amd64/kubectl -o /usr/local/bin/kubectl &&     chmod +x /usr/local/bin/kubectl &&     rm -rf /var/cache/apk/*" did not complete successfully: exit code: 3

<SNIP`>
```

**Comment by jge162** [comment]
Have you tried adding the proxy to daemon.json? Please edit proxy per your use case. 

```yaml
{
  "proxies": {
    "default": {
      "httpProxy": "http://proxy.behind_our_web_proxy.com:8080",
      "httpsProxy": "http://proxy.behind_our_web_proxy.com:8080"
    }
  }
}
```

**Reply by macoto-takahashi** [reply]
Ok, I'll try it.

**Reply by macoto-takahashi** [reply]
Docker daemon seems correct, but unfortunately result is the same.  
 
```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc., v2.16.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 23.0.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc version: v1.1.5-0-gf19387a
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.4.0-1087-kvm
 Operating System: Alpine Linux v3.17 (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.64GiB
 Name: example-runnerdeploy-mtmfk-5qj2z
 ID: 375dc52a-0f00-486e-bb1a-ec349c13ac51
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http://x.x.x.x:3128
 HTTPS Proxy: http://x.x.x.x:3128
 No Proxy: x.x.x.x/x
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```
I manually added ~/.docker/config.json and tried `docker build .`, but It couldn't work.  
When I tried `docker build . --build-arg="https_proxy=http://x.x.x.x:3128"` , I worked.

**Reply by macoto-takahashi** [reply]
Docker seems to pull FROM image in the Dockerfile successfully. I think Docker knows proxy.
But without --build-arg option, ` RUN ... apk add --no-cache --update openssl curl ca-certificates` in the Dockerfile cannot use proxy.

```
SNIP 

  #4 [1/3] FROM docker.io/library/alpine:3.17@sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126
  #4 sha256:42909a26917df41199aea555921cbce1f768cf03cbf4e1f29374960450546818
  #4 resolve docker.io/library/alpine:3.17@sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126 done
  #4 sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126 1.64kB / 1.64kB done
  #4 sha256:b6ca290b6b4cdcca5b3db3ffa338ee0285c11744b4a6abaa9627746ee3291d8d 528B / 528B done
  #4 sha256:9ed4aefc74f6792b5a804d1d146fe4b4a2299147b0f50eaf2b08435d7b38c27e 1.47kB / 1.47kB done
  #4 sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09 0B / 3.37MB 0.2s
  #4 sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09 3.37MB / 3.37MB 0.3s done
  #4 extracting sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09
  #4 extracting sha256:f56be85fc22e46face30e2c3de3f7fe7c15f8fd7c4e5add29d7f64b87abdaa09 0.1s done
  #4 DONE 0.4s
  
  #6 [2/3] COPY entrypoint.sh /entrypoint.sh
  #6 sha256:34ac351fa23d82b65df1d0e22532943346d6019527aa56ade19165dbda395ed1
  #6 DONE 0.0s
  
  #7 [3/3] RUN chmod +x /entrypoint.sh &&     apk add --no-cache --update openssl curl ca-certificates &&     curl -L https://storage.googleapis.com/kubernetes-release/release/v1.27.1/bin/linux/amd64/kubectl -o /usr/local/bin/kubectl &&     chmod +x /usr/local/bin/kubectl &&     rm -rf /var/cache/apk/*
  #7 sha256:ff17200[40](https://github.com/swfg1003727/kubespray/actions/runs/4990301507/jobs/8935322480#step:3:40)ab6185e26f0a5e170deb0f0b1e894f03a121d3cb85a2616f6b[43](https://github.com/swfg1003727/kubespray/actions/runs/4990301507/jobs/8935322480#step:3:43)a5b
  #7 0.212 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz
  #7 130.4 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/main: operation timed out
  #7 130.4 fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz
  #7 2[61](https://github.com/swfg1003727/kubespray/actions/runs/4990301507/jobs/8935322480#step:3:61).5 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.17/community: operation timed out
  #7 261.5 ERROR: unable to select packages:

SNIP
```

**Comment by ChristopherHX** [comment]
Because you are using a self-hosted runner, I suggest you to configure runner container hooks.

Then you can add your arguments of your environment to docker build here
https://github.com/actions/runner-container-hooks/blob/main/packages/docker/src/dockerCommands/container.ts#L153

After you build your change of the docker container hooks point `ACTIONS_RUNNER_CONTAINER_HOOKS` to your `index.js` file.

More details here https://github.com/actions/runner-container-hooks/blob/main/CONTRIBUTING.md, use the `docker` library.

For example I inject my self-signed certificate in all pulled containers, otherwise I have to include that details in every docker image I use locally.

**Reply by macoto-takahashi** [reply]
Thanks, 

Let me confirm one thing. 
The source code you introduced doesn't have "--build-env" option. this is a default runner behavior and default runner cannot handle the option when docker build, correct?

**Reply by ChristopherHX** [reply, marked as answer]
> The source code you introduced doesn't have "--build-env" option. this is a default runner behavior and default runner cannot handle the option when docker build, correct?

Yes the actions runner doesn't apply any arguments like `--build-env` or build arg from your job environment to docker build.

_In the past I have seen some even created a script called `docker`, which called docker itself with modified parameters._

I even think you can route the env variables of your job with container hooks as build arguments for docker build.

The runner container hooks docker is a official reimplementation of the integrated container management of actions/runner (however container hooks doesn't prevent you from using containers within docker in docker) 

By defining the environment variable `ACTIONS_RUNNER_CONTAINER_HOOKS` before executing the runner it will switch to container hooks for container management.

By using the container hooks, you can safely let the actions/runner autoupdate (I think for this to work you have to store the hook library file outside of the actions-runner folder) as long they are not breaking the existing container hooks interface.

**Reply by macoto-takahashi** [reply]
Got it. Thanks!
